platform/src/main/python/dlpx/virtualization/platform/migration_helper.py

```diff
@@ -0,0 +1,327 @@
+        Validates that the migration id is the correct type/format and then
+        return the canonical format of the id. Add the id as an array of
+        integers into the list of migration ids.
+        """
+        # First validate that the migration_id is the correct type/format.
+        self.__validate_migration_id(migration_id, impl_name)
+
+        # Then we must standardize the migration_id.
+        std_migration_id = self.__standardize_migration_id_to_array(
+            migration_id, impl_name)
+        std_string = '.'.join(str(i) for i in std_migration_id)
+
+        # Then we should check if this migration_id has already been used
+        if std_migration_id in self.__migration_ids:
+            raise MigrationIdAlreadyUsedError.fromMigrationId(
+                migration_id, std_string, impl_name)
+
+        # Lastly we should add this new array into the internal migration list.
+        self.__migration_ids.append(std_migration_id)
+
+        # Return back the standardized format of the migration id
+        return std_string
+
+    @staticmethod
+    def __validate_migration_id(migration_id, impl_name):
+        # First validate that the id is a string
+        if not isinstance(migration_id, basestring):
+            raise MigrationIdIncorrectTypeError(migration_id, impl_name)
+
+        # Next check if the id is the right format
+        if not PlatformUpgradeMigrations.MIGRATION_ID_REGEX.match(
+                migration_id):
+            raise MigrationIdIncorrectFormatError.from_fields(
+                migration_id, impl_name,
+                PlatformUpgradeMigrations.MIGRATION_ID_REGEX.pattern)
+
+    @staticmethod
+    def __standardize_migration_id_to_array(migration_id, impl_name):
+        # Split on the period and convert to integer
+        array = [int(i) for i in migration_id.split('.')]
+
+        #
+        # We cannot allow a migration id of essentially '0' because otherwise
+        # there would be no way to add a migration that goes before this.
+        #
+        if not any(array):
+            raise MigrationIdIncorrectFormatError(
+                "The migration id '{}' used in the function '{}' cannot be"
+                " used because a 0 migration id is not allowed.".format(
+                    migration_id, impl_name))
+
+        # Next we want to trim all trailing zeros so ex: 5.3.0.0 == 5.3
+        while array:
+            if not array[-1]:
+                # Remove the last element which is a zero from array
+                array.pop()
+            else:
+                break
+
+        return array
+
+    def get_sorted_ids(self):
+        # First sort the migration ids
+        self.__migration_ids.sort()
+
+        # Then convert all these arrays to the usual string format.
+        return [
+            '.'.join(str(i) for i in migration_id)
+            for migration_id in self.__migration_ids
+        ]
+
+    def get_repository_impls_to_exec(self, migration_id_list):
```

Comment:
It looks like these _get_xxx_to_exec methods are identical between the Lua and Platform classes. Should these go into the superclass?

Reply:
hmm they're technically different. One takes in a list of migraitons (platform) and the other takes in only one (lua)
Also the actual function they call do something different although I could just overwrite the private method.

Although since it's python we don't have to stick to the type (ie list version single string)...
Do you think it would be confusing to merge? what would I call the variable name?

Reply:
I didn't realize when I commented that the semantics were a little different. I'm totally fine leaving them separate.